### PR TITLE
Adding fully asynchronous versions of connect and publish.

### PR DESCRIPTION
**Pull request by stebet**
## Proposed Changes

Now that the library is using Pipelines we can create fully asynchronous versions of the RabbitMQ operations. Here is a draft PR that implements async methods to run through the connection handshake and makes an async version of BasicPublish.

TODO:
 - [ ] Run comparison benchmarks
 - [ ] Documentation
 - [ ] More tests

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [X] I have read the `CONTRIBUTING.md` document
- [X] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [X] All tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

Going to make this a draft for now. If this makes sense, we can gradually move other operations to be async as well, and hopefully even remove the synchronous versions completely later on.

### CODE REVIEW

**Comment by lukebakken** [comment]
Thanks @stebet! I don't want to have pull requests staying open for a long time so I'm going to fix conflicts here, review, and get it merged ASAP!

**Comment by PauloHMattos** [comment]
@stebet with this change will be possible to avoid the copy in BasicPublish as described [here](https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1184#discussioncomment-2499975)?

**Comment by stebet** [comment]
> @stebet with this change will be possible to avoid the copy in BasicPublish as described [here](https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1184#discussioncomment-2499975)?

I think so yes. Let me read over this again tomorrow to make sure.

**Comment by stebet** [comment]
> @stebet with this change will be possible to avoid the copy in BasicPublish as described [here](https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1184#discussioncomment-2499975)?

Yes, this PR solves this issue where memory usage goes up the roof if the producing app is sending messages faster than they can be delivered over the network, since the output buffer is no longer unlimited. It is now limited to 128 messages and if that is full, the `BasicPublish` method blocks until more space is available in the buffer.

**Comment by lukebakken** [comment]
I'm trying to suss out the CI build failures. Works on my machine, of course.

**Comment by stebet** [comment]
Anything left to resolve here @lukebakken?

**Comment by lukebakken** [comment]
@stebet nope, just waiting around to see if anyone else will review. I just got back from some time off. Merging it!

**Review by lukebakken — APPROVED** [review]

**Comment by lukebakken** [comment]
@stebet let me know what you plan to "async-ify" next (if anything, you've contributed _so much_ so far) and I will work on another section of the code.

**Review by Kuinox — COMMENTED** [review]
Hello,
Are you really sure about that?
It can cause deadlocks in some scenarios:
https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#warning-deadlocks

File: projects/RabbitMQ.Client/client/impl/Connection.cs [hunk 1]
```diff
@@ -402,7 +407,16 @@ internal void OnCallbackException(CallbackExceptionEventArgs args)
 
         internal void Write(ReadOnlyMemory<byte> memory)
         {
-            _frameHandler.Write(memory);
+            var task = _frameHandler.WriteAsync(memory);
+            if (!task.IsCompletedSuccessfully)
+            {
+                task.AsTask().GetAwaiter().GetResult();
```
[the remark this answers is not in the document]
Comment (lukebakken) [reply projects/RabbitMQ.Client/client/impl/Connection.cs:413]: We would appreciate a pull request with your ideas for improvements. Thank you.

**Pull request merged**